buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -33,73 +33,141 @@
 
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 public class TestClustersPlugin implements Plugin<Project> {
 
     private static final String LIST_TASK_NAME = "listTestClusters";
     private static final String NODE_EXTENSION_NAME = "testClusters";
+    public static final String PROEPRTY_TESTCLUSTERS_RUN_ONCE = "_testclusters_run_once";
 
     private final Logger logger =  Logging.getLogger(TestClustersPlugin.class);
 
+    // this is static because we need a single mapping across multi project builds, as some of the listeners we use,
+    // like task graph are singletons across multi project builds.
+    private static final Map<Task, List<ElasticsearchNode>> taskToCluster = new ConcurrentHashMap<>();
+
     @Override
     public void apply(Project project) {
-        NamedDomainObjectContainer<? extends ElasticsearchNode> container = project.container(
+
+        // enable the DSL to describe clusters
+        NamedDomainObjectContainer<ElasticsearchNode> container = createTestClustersContainerExtension(project);
+
+        // provide a task to be able to list defined clusters.
+        createListClustersTask(project, container);
+
+        // create DSL for tasks to mark clusters these use
+        creteUseClusterTaskExtension(project);
+
+        // There's a single Gradle instance for multi project builds, this means that some configuration needs to be
+        // done only once even if the plugin is applied multiple times as a part of multi project build
+        ExtraPropertiesExtension rootProperties = project.getRootProject().getExtensions().getExtraProperties();
+        if (rootProperties.has(PROEPRTY_TESTCLUSTERS_RUN_ONCE) == false) {
+            rootProperties.set(PROEPRTY_TESTCLUSTERS_RUN_ONCE, true);
+            // When running in the Daemon it's possible for this to hold references to past
+            taskToCluster.clear();
+
+            // When we know what tasks will run, we claim the clusters of those task to differentiate between clusters
+            // that are defined in the build script and the ones that will actually be used in this invocation of gradle
+            // we use this information to determine when the last task that required the cluster executed so that we can
+            // terminate the cluster right away and free up resources.
+            configureClaimClustersHook(project);
+
+            // Before each task, we determine if a cluster needs to be started for that task.
+            configureStartClustersHook(project);
+
+            // After each task we determine if there are clusters that are no longer needed.
+            configureStopClustersHook(project);
+        } else {
+            logger.lifecycle("Allready configured for {}", project);
```

Comment:
This seems verbose for a lifecycle message

Reply:
This is an oversight TBH, but logging is too verbose overall, which I plan to address in a different PR.

Reply:
Sorry, I did the logging changes I mentioned but missed this one.